### PR TITLE
Add options and option extensions

### PR DESCRIPTION
**Pull request by htdvisser**
ping @pseudomuto 

This pull request builds on top of https://github.com/pseudomuto/protokit/pull/2 and https://github.com/pseudomuto/protokit/pull/4 to add support for options and extended options in files, services, methods, enums, enum values, messages and fields.

It supports two builtin protobuf options:

- `deprecated` (as asked in https://github.com/pseudomuto/protoc-gen-doc/issues/233#issuecomment-432961525)
- `idempotency_level` (available on Methods)

and three external extensions:

- `validator.field` for [go-proto-validators](https://github.com/mwitkow/go-proto-validators).
- `validate.rules` for [protoc-gen-validate](https://github.com/lyft/protoc-gen-validate).
- `google.api.http` for [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway).

## Notes

Right now for `validator.field` we use reflection to transform the field validation rules, but it could be nice to generate the code for that from `vendor/github.com/mwitkow/go-proto-validators/validator.proto`.

I'm currently importing `github.com/pseudomuto/protokit/fixtures/extend.proto`, maybe it's cleaner to copy it.

### CODE REVIEW

**Comment by htdvisser** [comment]
It's not forgotten, I see it every time when I click on "Pull Requests" in Github's top bar :wink:

This PR is just waiting for @pseudomuto, and I suspect he's been a bit busy. From his Github activity I suspect that he just returned from a holiday, so maybe he'll find time for it soon.

Everything is basically done, so moving forward, we need to:

1. Tag a new version of protokit (@pseudomuto)
2. Update Gopkg.toml/Gopkg.lock in this PR (@htdvisser)
3. Review this PR (@pseudomuto)
4. Fix things that came out of review (@htdvisser)
5. Merge the PR (@pseudomuto)
6. Possibly tag a new version of protoc-gen-doc (@pseudomuto)

**Comment by igrayson** [comment]
The ability to emit field/service/method options in json output would be fantastically useful to us (currently using a [forked protobuf.js](https://github.com/dcodeIO/protobuf.js/pull/1127) to achieve this). Looking forward to progress!

**Comment by pseudomuto** [comment]
My apologies for the long-delayed response here. Between vacation, conferences, and work I haven't had enough free time to review this.

Good news is, I've got time this week. I'll make sure I've got this reviewed by the end of the week.

**Comment by htdvisser** [comment]
It didn't look like @pseudomuto reviewed this yet, so I took the liberty of adding support for the `validate.rules` option from [github.com/lyft/protoc-gen-validate](https://github.com/lyft/protoc-gen-validate) as well.

File: Makefile [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 
-EXAMPLE_DIR=$(shell pwd)/examples
+EXAMPLE_DIR=$(PWD)/examples
 DOCS_DIR=$(EXAMPLE_DIR)/doc
 PROTOS_DIR=$(EXAMPLE_DIR)/proto
 
-EXAMPLE_CMD=protoc --plugin=protoc-gen-doc -Iexamples/proto --doc_out=examples/doc
-DOCKER_CMD=docker run --rm -v $(DOCS_DIR):/out:rw -v $(PROTOS_DIR):/protos:ro -v $(EXAMPLE_DIR)/templates:/templates:ro pseudomuto/protoc-gen-doc
+EXAMPLE_CMD=protoc --plugin=protoc-gen-doc -Ivendor -Itmp/googleapis -Iexamples/proto --doc_out=examples/doc
+DOCKER_CMD=docker run --rm -v $(DOCS_DIR):/out:rw -v $(PROTOS_DIR):/protos:ro -v $(EXAMPLE_DIR)/templates:/templates:ro -v $(PWD)/vendor/github.com/mwitkow:/usr/local/include/github.com/mwitkow:ro -v $(PWD)/tmp/googleapis/google/api:/usr/local/include/google/api:ro pseudomuto/protoc-gen-doc
```
Comment (pseudomuto) [inline Makefile]: This needs `-v $(PWD)/vendor/github.com/lyft:/usr/local/include/github.com/lyft:ro` in order to run `make docker_test`

File: Gopkg.toml [hunk 2]
```diff
@@ -34,5 +35,5 @@
   version = "^1.1.0"
 
 [[constraint]]
+  branch = "master"
   name = "github.com/pseudomuto/protokit"
-  version = "0.1.0"
```
Comment (pseudomuto) [inline Gopkg.toml]: I'll release 0.2.0 today so this can be pinned to a release

File: extensions/extensions.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+// Package extensions implements a system for working with extended options.
+package extensions
```
Comment (pseudomuto) [inline extensions/extensions.go]: 👍

File: Gopkg.toml (hunk 2 above)
Comment (pseudomuto) [reply Gopkg.toml]: ^ done

**Comment by htdvisser** [comment]
merge

**Comment by pseudomuto** [comment]
Released! - https://github.com/pseudomuto/protoc-gen-doc/releases/tag/v1.3.0

Thanks again for the PR @htdvisser!

**Comment by rogchap** [comment]
This is a great addition thanks!... Sadly only outputs for the HTML template. 
Need to add the changes to the other templates... selfishly the Markdown template :D

**Comment by htdvisser** [comment]
@rogchap if you want to submit a pull request, feel free to get inspired by [the markdown template I wrote](https://github.com/TheThingsNetwork/lorawan-stack/blob/master/api/api.md.tmpl) for [TheThingsNetwork/lorawan-stack](https://github.com/TheThingsNetwork/lorawan-stack)

**Pull request merged**